syntax = "proto3"; 

package transport_catalogue_serialize;

message Coordinates {
	double lat = 1;
	double lng = 2;
}

message Stop {
	uint64 id = 1;
	bytes stop_name = 2;
	Coordinates coordinates = 3;
}

message StopsDistance{
	uint64 stop_from_id = 1;
	uint64 stop_to_id = 2;
	int32 distance = 3;
}

message Bus {
	string bus_name = 1;
	repeated uint64 stop_ids = 2;
	bool is_roundtrip = 3;
}

message Point {
	double x = 1;
	double y = 2;
}

message Rgb {
	uint32 red = 1;
	uint32 green = 2;
	uint32 blue = 3;
}

message Rgba {
	uint32 red = 1;
	uint32 green = 2;
	uint32 blue = 3;
	double opacity = 4;
}


message Color {
  oneof value {
    Rgb rgb = 1;
	Rgba rgba = 2;
	string color_name = 3;
	bool is_monostate = 4;
  }
}

message RenderSettings {
	double width = 1;
	double height = 2;
	double padding = 3;
	double line_width = 4;
	double stop_radius = 5;
	int32 bus_label_font_size = 6;
	Point bus_label_offset = 7;
	int32 stop_label_font_size = 8;
	Point stop_label_offset = 9;
	Color underlayer_color = 10;
	double underlayer_width = 11;
	repeated Color color_palette = 12;
}

message RoutingSettings {
	double bus_wait_time = 1;
	double bus_velocity = 2;
}

message TransportCatalogue {
	repeated Stop stop = 1;
	repeated StopsDistance distance = 2;
	repeated Bus bus = 3;
	RenderSettings render_settings = 4;
	RoutingSettings routing_settings = 5;
}